rmf_websocket/src/rmf_websocket/BroadcastClient.cpp

```diff
@@ -15,216 +15,151 @@
-              c->_uri.c_str());
-            c->_connected = true;
-          }
-
-          std::unique_lock<std::mutex> lock(c->_wait_mutex);
-          c->_cv.wait(lock,
-          [c]()
-          {
-            return !c->_queue.empty();
-          });
+          this->_processing_thread();
+        });
+  }
 
-          while (!c->_queue.empty())
-          {
-            std::lock_guard<std::mutex> lock(c->_queue_mutex);
-            websocketpp::lib::error_code ec;
-            const std::string& msg = c->_queue.front().dump();
-            c->_client.send(c->_hdl, msg, websocketpp::frame::opcode::text, ec);
-            if (ec)
-            {
-              RCLCPP_ERROR(
-                c->_node->get_logger(),
-                "BroadcastClient unable to publish message: %s",
-                ec.message().c_str());
-              // TODO(YV): Check if we should re-connect to server
-              break;
-            }
-            c->_queue.pop();
-          }
-        }
-      });
+  void log(const std::string& str)
+  {
+    RCLCPP_ERROR(
+      this->_node->get_logger(),
+      "%s",
+      str.c_str()
+    );
   }
 
-  // Publish a single message
+  //============================================================================
   void publish(const nlohmann::json& msg)
   {
-    std::lock_guard<std::mutex> lock(_queue_mutex);
+    /// _queue is thread safe. No need to lock.
     _queue.push(msg);
-    _cv.notify_all();
   }
 
-  // Publish a vector of messages
+  //============================================================================
   void publish(const std::vector<nlohmann::json>& msgs)
   {
-    std::lock_guard<std::mutex> lock(_queue_mutex);
-    for (const auto& msg : msgs)
+    for (auto msg: msgs)
+    {
+      /// _queue is thread safe. No need to lock.
       _queue.push(msg);
-    _cv.notify_all();
+    }
   }
 
+  //============================================================================
   void set_queue_limit(std::optional<std::size_t> limit)
   {
-    std::lock_guard<std::mutex> lock(_queue_mutex);
-    _queue_limit = limit;
+    /// _queue is thread safe. No need to lock.
+    if (limit.has_value())
+      _queue.resize(limit.value());
   }
 
+  //============================================================================
   ~Implementation()
   {
-    _shutdown = true;
-    if (_processing_thread.joinable())
+    _stop = true;
+    _endpoint.interrupt_waits();
+    _consumer_thread.join();
+  }
+
+private:
+  //============================================================================
+  /// Background consumer thread
+  void _processing_thread()
+  {
+    _endpoint.connect();
+    _check_conn_status_and_send(std::nullopt);
+    while (!_stop)
     {
-      _processing_thread.join();
+
+      auto item = _queue.pop_item();
+      if (!item.has_value())
+      {
+        _queue.wait_for_message();
+        continue;
+      }
+      _check_conn_status_and_send(item);
```

Comment:
I think we can avoid any threading/mutexes in `ClientWebSocketEndpoint` by running the event loop here and using `dispatch` https://www.boost.org/doc/libs/1_76_0/boost/asio/dispatch.hpp to run an action that pops an item from the queue and call `_check_conn_status_and_send`.

Reply:
[6ab662f](https://github.com/open-rmf/rmf_ros2/pull/329/commits/6ab662fceea44ec77bba66993c926f6f77fa7f2b)